aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -2136,5 +2138,31 @@ Tensor sum_coo(const Tensor &self, c10::optional<ScalarType> dtype) {
   return self._values().sum(dtype);
 }
 
+Tensor sum_sparse_coo(const Tensor& self, at::OptionalIntArrayRef dim, bool keepdim, c10::optional<ScalarType> dtype) {
+  Tensor result;
+  if (dim.has_value()) {
+    if (dtype.has_value()) {
+      result = at::_sparse_sum(self, *dim, *dtype);
+    } else {
+      if (c10::isIntegralType(self.scalar_type(), true)) {
+        result = at::_sparse_sum(self, *dim, at::kLong);
+      } else {
+        result = at::_sparse_sum(self, *dim);
+      }
+    }
+  } else {
+    result = sum_coo(self, dtype);
+  }
+  if (keepdim) {
+    auto dim_mask = make_dim_mask(dim, self.dim());
+    for (int dim = 0; dim < self.dim(); dim++) {
```

Comment:
should these unsqueeze start from the last dimension instead of the first ones?

Reply:
no, because dim_mask and the dim loop correspond to the `self` shape and unsqueeze inserts (read: restores) the removed dimensions at the position that is in the reference of `self`, not of `result`, shape.

Consider the following example:
```python
>>> mask_dim = [True, True, False]
>>> dim = tuple(i for i in range(len(mask_dim)) if mask_dim[i])
>>> dim
(0, 1)
>>> # Using algorithm in this PR:
>>> result = a.sum(dim=dims, keepdim=False)
>>> for i in range(a.ndim):
...     result = result.unsqueeze(i) if mask_dim[i] else result
... 
>>> result
tensor([[[16, 20]]])
>>> # Using unsqueeze starting from the last dimension:
>>> result = a.sum(dim=dims, keepdim=False)
>>> for i in reversed(range(a.ndim)):
...     result = result.unsqueeze(i) if mask_dim[i] else result
... 
>>> result
tensor([[[16],
         [20]]])
>>> # The expected result:
>>> a.sum(dim=dim, keepdim=True)
tensor([[[16, 20]]])
```
As you can see, the algorithm implemented in this PR gives the expected result while reversing the order of unsqueeze not.